ENH: Add the linalg extension to the array_api submodule

asmeurer opened the pull request:
This is an in progress PR for adding the [linalg extension](https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html) to the array API submodule.

asmeurer committed:
Add functions from the linear algebra array API extension

asmeurer committed:
Move linear algebra named tuples to the linalg file to avoid circular imports

asmeurer committed:
Fix discrepancies between the NumPy cross() and the array API cross()

asmeurer committed:
Fix some bugs in the array API linear algebra module

asmeurer committed:
Fix some more bugs in the array API linear algebra module

asmeurer committed:
Update the array_api linalg extension code for changes to the array_api submodule

asmeurer committed:
Fix Array class rename in the array API linear algebra extension

asmeurer committed:
Remove references to eig and eigvals in the array API submodule

These functions won't be added until next year so there is no reason to have
commented out versions of them sitting around.

asmeurer committed:
Remove a todo item from the numpy/_array_api/__init__.py docstring

asmeurer committed:
Remove numpy/array_api/_linear_algebra_functions.py

The functions are now all in numpy/array_api/linalg.py, and the public ones
(in the top-level namespace as opposed to the numpy.array_api.linalg extension
subnamespace) are differentiated by which names are included in
numpy.array_api.__all__.

asmeurer commented:
Ugh, I didn't know that @charris was going to do a squash merge of my other PR. Does anyone know how to sanely rebase this so that it keeps my existing merge conflict resolutions from [`aeced8f` (#19980)](https://github.com/numpy/numpy/pull/19980/commits/aeced8f21d044eb7115b9aed2634324ece05d1d9)?

asmeurer committed:
Fix the mT attribute in the array_api extension

asmeurer committed:
Use mT in the array_api cholesky

asmeurer committed:
Update the array API linalg extension signatures

norm() has been split into the matrix_norm() and vector_norm() functions.
These functions are not yet updated to function properly.

asmeurer committed:
Update the array_api linalg extension namedtuple names

BvB93 commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,346 @@
+
+# Note: the name here is different from norm(). The array API norm is split
+# into matrix_norm and vector_norm().
+def vector_norm(x: Array, /, *, axis: Tuple[int, int] = (-2, -1), keepdims: bool = False, ord: Optional[Union[int, float, Literal[np.inf, -np.inf, 'fro', 'nuc']]] = 'fro') -> Array:
```
```suggestion
def vector_norm(x: Array, /, *, axis: Tuple[int, int] = (-2, -1), keepdims: bool = False, ord: Optional[Union[int, float, Literal['fro', 'nuc']]] = 'fro') -> Array:
```
Literal floats aren't supported. [PEP 586](https://www.python.org/dev/peps/pep-0586/#illegal-parameters-for-literal-at-type-check-time) does mention potential support for literal floats in the future, but there have been no follow-up PEPs as of yet.

The same comment applies to `matrix_norm`.

BvB93 replied:
As a reminder: the `Literal` issue still applies to `matrix_norm` & `vector_norm` and should be resolved.

rgommers replied:
Changed to `Optional[Union[int, float]]` in the meantime, and the spec removed the string versions of the allowed inputs.

seberg reviewed (commented):
Probably, should just keep things moving here (as soon as the rebase is cleared up).  But looking through, I had some questions/comments, partly about the API standard, but maybe you can forward them there if necessary, since I could not find a discussion for a few of them.

For the rebase, maybe it is easier to just squash all of your commits and rebase after that?  That way the conflicts are all in one go and don't keep recurring at least and there is probably little reason to preserve the history here anyway.

In my opinion, you could probably remove most (probably not all) of the `Note:` comments (or shorten them). It is clear that the code means there is a difference to NumPy proper, in some cases it is unclear/interesting what the difference is though.

BvB93 commented on numpy/array_api/_array_object.py:
```diff
@@ -391,6 +394,8 @@ def __bool__(self: Array, /) -> bool:
         # Note: This is an error here.
         if self._array.ndim != 0:
             raise TypeError("bool is only allowed on arrays with 0 dimensions")
+        if self.dtype not in _boolean_dtypes:
+            raise ValueError("bool is only allowed on boolean arrays")
```
Is there a reason why invalid dtypes sometimes raise a `ValueError` while in others they raise a `TypeError`?

asmeurer replied:
Likely I changed it at some point in my previous PR but haven't yet backported the change to this one (I originally started this as part of https://github.com/numpy/numpy/pull/19739 but moved it to a separate branch). Thanks for the heads up.

rgommers replied:
I checked this, there are indeed still 2 places where `ValueError` is used: in `__float__` and in `__bool__`. Let's fix that.

rgommers replied:
Actually, @asmeurer you seem to have done this on purpose:
```python

def test_python_scalar_construtors():
    b = asarray(False)
    i = asarray(0)
    f = asarray(0.0)

    assert bool(b) == False
    assert int(i) == 0
    assert float(f) == 0.0
    assert operator.index(i) == 0

    # bool/int/float should only be allowed on 0-D arrays.
    assert_raises(TypeError, lambda: bool(asarray([False])))
    assert_raises(TypeError, lambda: int(asarray([0])))
    assert_raises(TypeError, lambda: float(asarray([0.0])))
    assert_raises(TypeError, lambda: operator.index(asarray([0])))

    # bool/int/float should only be allowed on arrays of the corresponding
    # dtype
    assert_raises(ValueError, lambda: bool(i))
    assert_raises(ValueError, lambda: bool(f))

    assert_raises(ValueError, lambda: int(b))
    assert_raises(ValueError, lambda: int(f))

    assert_raises(ValueError, lambda: float(b))
    assert_raises(ValueError, lambda: float(i))

    assert_raises(TypeError, lambda: operator.index(b))
    assert_raises(TypeError, lambda: operator.index(f))
```
Can you comment?

BvB93 commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,346 @@
+
+# Note: svdvals is not in NumPy (but it is in SciPy). It is equivalent to
+# np.linalg.svd(compute_uv=False).
+def svdvals(x: Array, /) -> Union[Array, Tuple[Array, ...]]:
+    return Array._new(np.linalg.svd(x._array, compute_uv=False))
```
```suggestion
def svdvals(x: Array, /) -> Array:
    return Array._new(np.linalg.svd(x._array, compute_uv=False))
```
This expression will never return a tuple of arrays.
I imagine this might be a leftover from an earlier version of the spec, one wherein `svd`'s keyword arguments were exposed?

asmeurer replied:
Yes, I think that was the case. I'll add this to https://github.com/data-apis/array-api/pull/277 as well.

BvB93 commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,346 @@
+import numpy.linalg
+import numpy as np
+
+class eighresult(NamedTuple):
+    eigenvalues: Array
+    eigenvectors: Array
+
+class qrresult(NamedTuple):
+    q: Array
+    r: Array
+
+class slogdetresult(NamedTuple):
+    sign: Array
+    logabsdet: Array
+
+class svdresult(NamedTuple):
+    u: Array
+    s: Array
+    vh: Array
```
Is there any reason for not going with camel-case names here?

asmeurer replied:
The field names are part of the spec. If you think those should be changed you should bring it up in the spec repo. The class name was chosen by me. We can change it to something else if you feel that would be better.

BvB93 replied:
Ah yes, to clarify: my comment was aimed at the class names, not the fields names. 
I was aiming at something like `SVDResult` instead of `svdresult` for example.

asmeurer replied:
Yes, we can change that. I've seen lowercase convention used sometimes for namedtuples, but I think there's not a consistent pattern. Even in NumPy itself it seems like both conventions are used in some places.

asmeurer commented:
@seberg most of your comments are actually about the code from the already-merged https://github.com/numpy/numpy/pull/19937, and they will disappear once I rebase this branch properly.

asmeurer commented on numpy/array_api/_array_object.py:
```diff
@@ -488,9 +495,18 @@ def __int__(self: Array, /) -> int:
         if self._array.ndim != 0:
             raise TypeError("int is only allowed on arrays with 0 dimensions")
+        if self.dtype not in _integer_dtypes:
+            raise ValueError("int is only allowed on integer arrays")
```
I would bring these issues up in the spec repo. I expect this was done to be consistent with the rest of the spec, which generally disallows (or at best, doesn't require) cross-kind casting.

asmeurer commented on numpy/array_api/_array_object.py:
```diff
@@ -992,6 +1013,12 @@ def dtype(self) -> Dtype:
 
+    # Note: mT is new in array API spec (see matrix_transpose)
+    @property
+    def mT(self) -> Array:
```
The new APIs for matrix transpose were chosen so that they can be as backwards compatible as possible with libraries like NumPy that already use .T for reversing the axes. See https://github.com/data-apis/array-api/issues/228 and the corresponding PRs.

rgommers replied:
> I feel this was discussed in NumPy before (including the potential `.T` limitation!), time to revive that discussion!? 

Yes I think so.

> I find it mildly annoying that we do these things in the new namespace, but seem avers to discussing it for NumPy proper.

I don't think we are, it's just that no one followed up/through on making `matrix_transpose` public, and the discussion was framed around a big backcompat break in `.T`, which did not help. I think no one would object to making the private `matrix_transpose` public, it just needs doing?

> If we can get away with doing these things in NumPy it would be the best way to move code towards the better standard!

agreed

asmeurer commented on numpy/array_api/_statistical_functions.py:
```diff
@@ -24,6 +35,8 @@ def mean(
 ) -> Array:
+    if x.dtype not in _floating_dtypes:
+        raise TypeError("Only floating-point dtypes are allowed in mean")
     return Array._new(np.mean(x._array, axis=axis, keepdims=keepdims))
```
These three were also discussed at https://github.com/data-apis/array-api/issues/202. I'm not following why these functions should upcast, however. Unlike sum and prod, the mean of a set of numbers does not exceed its maximum, so it cannot overflow. The internal calculation might overflow if implemented in a certain way, but the spec says nothing about how these functions should be calculated internally. I may be missing something though, so if I am please raise this on the array-apis repo.

seberg replied:
I have scanned that discussion, and I think the meaty stuff was likely discussed at meetings? I don't think there is a lot about the choice for *float* reductions (e.g. bfloat16/float32)?  For one it includes this comment: https://github.com/data-apis/array-api/issues/202#issuecomment-863402982  which is *not* what is implemented here (or in the current version of the standard?).
I am happy to explain why I am curious about the choice (see below), but since I don't have the full picture (and can't make it easily), I will just have to be content with you handling it...

I probably misunderstood [this comment](https://github.com/data-apis/array-api/pull/238#issuecomment-922135271) and you are actually worried about summing many integer valued floats and getting an *exact* integral result!?  Rather than overflow/catastrophic annihilation during the summation?

To me it seemed that the main argument would be *precision*, i.e. summing too many `float32`, and especially `bfloat16` naively quickly runs into catastrophic annihilation and bogus results.  But `mean`, `var`, `std` will all run into the problem just the way `sum` does:
```
arr_float32 = np.ones(10**9).astype(dtype=np.float32).reshape(-1, 2)
arr_float32.sum(axis=0, dtype=np.float64)  / arr_float32.shape[0]
# array([1., 1.])   -- dtype=... since array API would do that implicitly
arr_float32.mean()
# array([0.03355443, 0.03355443], dtype=float32)
```
(NumPy uses a more stable summation without the reshape, which makes showing this a bit harder...)

rgommers replied:
The specification only specifies the dtype of the _return value_. What dtype is used for _intermediate results in the calculation_ is unspecified, and completely up to the library. As is the algorithm used. So it'd be perfectly valid to upcast to `int64` / `float64`, do the calculations, then downcast (and this is a common pattern for low-precision dtypes).

`sum` and `prod` can have outputs that overflow (mathematically), while `mean`, `min`, `max` etc. cannot. Furthermore this regularly happens in practice, for example scikit-image has code to take the sum over all axis for images, and those are typically ~1000x1000 px. Hence the different treatment of `sum` and `prod`.

rgommers replied:
This code is also no longer present in the PR diff, so I'll resolve the comment.

rgommers replied:
Now that I'm touching this comment, I see why it's confusing - I'll edit it to make clear it's about the returned array.

asmeurer commented on numpy/array_api/_statistical_functions.py:
```diff
@@ -64,8 +88,15 @@ def sum(
+    # Note: sum() and prod() always upcast float32 to float64 for dtype=None
+    # We need to do so here before summing to avoid overflow
+    if dtype is None and x.dtype == float32:
+        x = asarray(x, dtype=float64)
```
hmm, not sure why I didn't use the dtype parameter of sum(). I guess I was confused by it not having the same behavior with dtype=None, but yeah, we can still just pass it in manually and np.sum will do the right thing.

rgommers replied:
Also still needs addressing, taking this along.

rgommers replied:
Also will add a comment that `np.sum` does already do the right thing here for integers:
```
In [3]: x = np.arange(5, dtype=np.int8)

In [4]: np.sum(x, dtype=None)
Out[4]: 10

In [5]: np.sum(x, dtype=None).dtype
Out[5]: dtype('int64')

In [6]: x = np.arange(5, dtype=np.int32)

In [7]: np.sum(x, dtype=None).dtype
Out[7]: dtype('int64')
```

asmeurer commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,346 @@
+def svdvals(x: Array, /) -> Union[Array, Tuple[Array, ...]]:
+    return Array._new(np.linalg.svd(x._array, compute_uv=False))
+
+# Note: axes must be a tuple, unlike np.tensordot where it can be an array or array-like.
```
The array API namespace doesn't really have a notion of "array-like". The point here was more about the accepting of array than array-like. np.tensordot does appear to accept an array for the axes parameter. I'm also probably not going to add a restrictive type check here, a with the rest of the module. This is mostly just a note about the type annotation and what types are actually supported by the spec.

rgommers replied:
The actual code is correct, and the discussion is about two words in a corner case of a comment. Response seems okay, so I'll resolve this.

asmeurer commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,346 @@
+# Note: vecdot is not in NumPy
+def vecdot(x1: Array, x2: Array, /, *, axis: Optional[int] = None) -> Array:
+    if axis is None:
+        axis = -1
```
Yeah, it seems odd to me too. Probably just an oversight, given that this is a new function.

kgryte replied:
This will be updated in the array API specification (see https://github.com/data-apis/array-api/pull/281). Thanks for flagging.

rgommers replied:
This still needs updating to `-1`. Spec was changed to address @seberg's comment.

asmeurer committed:
Merge branch 'array-api-updates1' into array-api-linalg

asmeurer committed:
Fix the signature for the array API vector_norm

asmeurer committed:
Fix the implementation of the array API matrix_norm() and vector_norm()

asmeurer committed:
Fix the logic in vector_norm()

asmeurer committed:
Use CamelCase for the array_api linalg namedtuple classes

asmeurer commented:
I fixed the git history (for anyone wondering, I rewound the branch pre-merge, redid the merge against main, then cherry-picked the post-merge commits).

I would appreciate more reviews on this, but I don't think it should be merged yet. The reason is that it is seriously undertested. The array API test suite has almost no tests for the linalg extension, except for very basic checks. So I would like to do some work there before I feel confident about this, especially since there are a number of differences from the base numpy.linalg functions.

asmeurer committed:
Add notes for array API linalg functions that aren't in np.linalg

asmeurer committed:
Fix the array API diagonal and trace implementations

The NumPy versions operate on the first two axes, but the array API diagonal
should operate on the last two, so that they work correctly on stacks of
matrices. See https://github.com/data-apis/array-api/pull/241.

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+        raise TypeError('Only floating-point dtypes are allowed in cholesky')
+    L = np.linalg.cholesky(x._array)
+    if upper:
+        return Array._new(L).mT
```
This should be `.mH` to be correct for complex inputs, but there is no support for complex numbers in Array API, right?

kgryte replied:
No support for complex numbers, yet. Complex number support should be added in the 2022 revision.

asmeurer replied:
Yes, once complex numbers are added next year, several functions in the namespace will need to be updated, including several in linalg.

rgommers replied:
Answered, so resolving.

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+    # dimensional arrays.
+    if x.ndim < 2:
+        raise np.linalg.LinAlgError("1-dimensional array given. Array must be at least two-dimensional")
+    S = np.linalg.svd(x._array, compute_uv=False)
```
This could be a call to `numpy.array_api.svdvals`.

rgommers replied:
That would require more work, because then `S` is an `Array` object, so more unpacking/repacking is needed further down.

This is correct as is, so I'll resolve the cocmment.

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+
+    See its docstring for more information.
+    """
+    # Note: qr is supposed to support stacks of matrices, but
+    # np.linalg.qr does not yet.
```
Should be working since https://github.com/numpy/numpy/pull/19151

asmeurer replied:
Yep, can confirm from the test suite that this is working. This comment must be a leftover from the initial implementation.

rgommers replied:
This still needs addressing.

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+# Note: the keyword argument name upper is different from np.linalg.eigh
+def eigh(x: Array, /) -> EIGHResult:
+    """
+    Array API compatible wrapper for :py:func:`np.eig <numpy.eigh>`.
```
```suggestion
    Array API compatible wrapper for :py:func:`np.linalg.eigh <numpy.linalg.eigh>`.
```

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+# Note: the keyword argument name upper is different from np.linalg.eigvalsh
+def eigvalsh(x: Array, /) -> Array:
+    """
+    Array API compatible wrapper for :py:func:`np.eigvalsh <numpy.eigvalsh>`.
```
```suggestion
    Array API compatible wrapper for :py:func:`np.linalg.eigvalsh <numpy.linalg.eigvalsh>`.
```

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+    # Note: the restriction to floating-point dtypes only is different from
+    # np.linalg.norm.
+    if x.dtype not in _floating_dtypes:
+        raise TypeError('Only floating-point dtypes are allowed in norm')
```
```suggestion
        raise TypeError('Only floating-point dtypes are allowed in matrix_norm')
```

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+# Note: the keyword argument name rtol is different from np.linalg.pinv
+def pinv(x: Array, /, *, rtol: Optional[Union[float, Array]] = None) -> Array:
+    """
+    Array API compatible wrapper for :py:func:`np.pinv <numpy.pinv>`.
```
```suggestion
    Array API compatible wrapper for :py:func:`np.linalg.pinv <numpy.linalg.pinv>`.
```

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+
+def qr(x: Array, /, *, mode: Literal['reduced', 'complete'] = 'reduced') -> QRResult:
+    """
+    Array API compatible wrapper for :py:func:`np.qr <numpy.qr>`.
```
```suggestion
    Array API compatible wrapper for :py:func:`np.linalg.qr <numpy.linalg.qr>`.
```

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+
+def slogdet(x: Array, /) -> SLOGDETResult:
+    """
+    Array API compatible wrapper for :py:func:`np.slogdet <numpy.slogdet>`.
```
```suggestion
    Array API compatible wrapper for :py:func:`np.linalg.slogdet <numpy.linalg.slogdet>`.
```

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+
+def solve(x1: Array, x2: Array, /) -> Array:
+    """
+    Array API compatible wrapper for :py:func:`np.solve <numpy.solve>`.
```
```suggestion
    Array API compatible wrapper for :py:func:`np.linalg.solve <numpy.linalg.solve>`.
```

IvanYashchuk commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+
+def svd(x: Array, /, *, full_matrices: bool = True) -> SVDResult:
+    """
+    Array API compatible wrapper for :py:func:`np.svd <numpy.svd>`.
```
```suggestion
    Array API compatible wrapper for :py:func:`np.linalg.svd <numpy.linalg.svd>`.
```

kgryte commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+
+# Note: the name here is different from norm(). The array API norm is split
+# into matrix_norm and vector_norm().
+def vector_norm(x: Array, /, *, axis: Optional[Union[int, Tuple[int, int]]] = None, keepdims: bool = False, ord: Optional[Union[int, float, Literal[np.inf, -np.inf]]] = 2) -> Array:
```
This followed PyTorch: [`tensordot`](https://pytorch.org/docs/stable/generated/torch.tensordot.html) uses `dims` and [`vector_norm`](https://pytorch.org/docs/stable/generated/torch.linalg.vector_norm.html) (and [`matrix_norm`](https://pytorch.org/docs/stable/generated/torch.linalg.matrix_norm.html)) uses `dim`.

Admittedly, this is, at first blush, somewhat inconsistent. Typically, we've tried to use the singular for when a single axis can be specified (e.g., see the [manipulation functions](https://github.com/data-apis/array-api/blob/f267ebd3505bc13d9f49569021669b6742bcd046/spec/API_specification/manipulation_functions.md), which have some overlap in terms of `axis` usage).

We did not do that for `tensordot` as the `axes` keyword does not behave the same as the other APIs. When an `int`, axes specifies the number of axes over which to contract, not the axis index.

asmeurer replied:
IMO matrix_norm should use axes and not axis. Although I didn't realize the manipulation functions do this too, so maybe it's not as inconsistent as it seems.

Actually, I'm not even sure why matrix_norm has an axis argument at all. It's the only function here that lets you specify the axes that matrices are stacked on. This is done for some vector APIs (`tensordot`, `cross`, and `vector_norm`), but none of the matrix APIs.

kgryte replied:
The `axis` kwarg for `matrix_norm` follows PyTorch which is the only library which had implemented this API prior to standardization. We could remove from the specification, making the default behavior the last two axes. PyTorch could still support the `axis` kwarg without being non-compliant.

asmeurer replied:
It also comes from `np.linalg.norm`. It was probably implemented there for consistency for the vector case, but IMO it makes less sense now that the function is split since no other matrix function does this (unless we want to add `axis` to the other matrix functions).

rgommers replied:
`axis` keyword to `matrix_norm` was removed (xref https://github.com/data-apis/array-api/issues/318).

`tensordot` was reviewed and kept as "int or tuple of 2 sequences".

kgryte commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+    eigenvectors: Array
+
+class QRResult(NamedTuple):
+    q: Array
```
This has been updated in the array API specification: https://github.com/data-apis/array-api/pull/313.

asmeurer committed:
Small fixes to numpy.array_api.linalg docstrings

Co-authored-by: Ivan Yashchuk <IvanYashchuk@users.noreply.github.com>

asmeurer committed:
Update numpy.array_api.linalg.vecdot to not accept axis=None

asmeurer committed:
Remove the axis argument from numpy.array_api.matrix_norm

asmeurer committed:
Make numpy.array_api.linalg.solve only allow stacks of matrices

asmeurer committed:
Update Literal types to not use floating-point literals

asmeurer committed:
Update the field names for the array API qr() and svd() namedtuples

asmeurer commented:
I have addressed all the review commands and updated all the latest changes from the spec. This is now ready for review.

rgommers committed:
Merge remote-tracking branch 'upstream/main' into array-api-linalg

rgommers committed:
MAINT: add `array_api.linalg` to public API test

rgommers commented on numpy/array_api/linalg.py:
```diff
@@ -0,0 +1,371 @@
+import numpy.linalg
+import numpy as np
+
+class EIGHResult(NamedTuple):
```
This still needs addressing.

rgommers committed:
MAINT: address all remaining review comments

rgommers reviewed (approved):
I went through the code again, and also reviewed all unresolved review comments in detail (fixing those up if needed). It all LGTM now, and CI is green except some "line too long" complaints which are best ignored.

So in it goes. Thanks @asmeurer & all reviewers!
